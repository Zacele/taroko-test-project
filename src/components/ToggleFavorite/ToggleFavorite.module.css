.button {
  padding: 8px 16px;
  cursor: pointer;
  background-color: #1976d2;
  color: white;
  border: none;
  height: 35px;
  margin-right: 8px;
  width: 100%;
  @media (min-width: 768px) {
    width: unset;
  }
}
