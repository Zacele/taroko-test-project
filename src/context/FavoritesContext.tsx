'use client';
import React, {
  createContext,
  useContext,
  useState,
  ReactNode,
  useEffect,
} from 'react';
import Cookies from 'js-cookie';

type FavoritesContextType = {
  favorites: number[];
  toggleFavorite: (id: number) => void;
};

const FavoritesContext = createContext<FavoritesContextType | undefined>(
  undefined,
);

type FavoritesProviderProps = {
  children: ReactNode;
};

const FAVORITES_COOKIE_NAME = 'favorites';

export const FavoritesProvider: React.FC<FavoritesProviderProps> = ({
  children,
}) => {
  // Initialize state with empty array or from cookies
  const [favorites, setFavorites] = useState<number[]>(() => {
    const cookieValue = Cookies.get(FAVORITES_COOKIE_NAME);
    return cookieValue ? JSON.parse(cookieValue) : [];
  });

  useEffect(() => {
    // Update cookie whenever favorites change
    Cookies.set(FAVORITES_COOKIE_NAME, JSON.stringify(favorites), {
      expires: 7,
    });
  }, [favorites]);

  const toggleFavorite = (id: number) => {
    setFavorites((currentFavorites) => {
      const isFavorite = currentFavorites.includes(id);
      const newFavorites = isFavorite
        ? currentFavorites.filter((favId) => favId !== id)
        : [...currentFavorites, id];

      return newFavorites;
    });
  };

  return (
    <FavoritesContext.Provider value={{ favorites, toggleFavorite }}>
      {children}
    </FavoritesContext.Provider>
  );
};

export const useFavorites = (): FavoritesContextType => {
  const context = useContext(FavoritesContext);
  if (context === undefined) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};
