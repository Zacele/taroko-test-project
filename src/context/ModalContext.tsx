'use client';
import { ModalState } from '@/interface';
import React, { createContext, useContext, useState, ReactNode } from 'react';

const ModalContext = createContext<ModalState | undefined>(undefined);

export const useModal = () => {
  const context = useContext(ModalContext);
  if (context === undefined) {
    throw new Error('useModal must be used within a ModalProvider');
  }
  return context;
};

export const ModalProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [open, setOpen] = useState(false);
  const [isEditMode, setIsEditMode] = useState(false);
  const [selectedId, setSelectedId] = useState<number | null>(null);

  const closeModal = () => {
    setOpen(false);
    setIsEditMode(false);
    setSelectedId(null);
  };

  const openCreateModal = () => {
    setOpen(true);
  };

  const openEditModal = (id: number) => {
    setIsEditMode(true);
    setOpen(true);
    setSelectedId(id);
  };

  return (
    <ModalContext.Provider
      value={{
        isOpen: open,
        isEdit: isEditMode,
        openModal: openCreateModal,
        closeModal,
        openEditModal,
        selectedId,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};
