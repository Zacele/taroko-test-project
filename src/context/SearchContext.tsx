'use client';
import { ContactType } from '@/interface';
import React, { createContext, useContext, ReactNode, useState } from 'react';

interface SearchContextType {
  searchQuery: string;
  setSearchQuery: (query: string) => void;
  filterContacts: (contacts: ContactType[]) => ContactType[];
}

const SearchContext = createContext<SearchContextType | undefined>(undefined);

interface SearchProviderProps {
  children: ReactNode;
}

export const useSearch = () => {
  const context = useContext(SearchContext);
  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }
  return context;
};

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [searchQuery, setSearchQuery] = useState('');

  const filterContacts = (contacts: ContactType[]): ContactType[] => {
    return contacts.filter((contact) =>
      `${contact.first_name} ${contact.last_name}`
        .toLowerCase()
        .includes(searchQuery.toLowerCase()),
    );
  };

  return (
    <SearchContext.Provider
      value={{ searchQuery, setSearchQuery, filterContacts }}
    >
      {children}
    </SearchContext.Provider>
  );
};
