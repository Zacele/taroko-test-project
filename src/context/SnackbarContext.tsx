'use client';
import React, { createContext, useContext, useState, ReactNode } from 'react';
import Snackbar from '@mui/material/Snackbar';
import Alert, { AlertColor } from '@mui/material/Alert';

type SnackbarContextType = {
  showSnackbar: (message: string, severity: AlertColor) => void;
};

const SnackbarContext = createContext<SnackbarContextType | undefined>(
  undefined,
);

type Props = {
  children: ReactNode;
};

export const SnackbarProvider: React.FC<Props> = ({ children }) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState('');
  const [severity, setSeverity] = useState<AlertColor>('info');

  const showSnackbar = (message: string, severity: AlertColor = 'info') => {
    setMessage(message);
    setSeverity(severity);
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <SnackbarContext.Provider value={{ showSnackbar }}>
      {children}
      <Snackbar open={open} autoHideDuration={2000} onClose={handleClose}>
        <Alert onClose={handleClose} severity={severity} sx={{ width: '100%' }}>
          {message}
        </Alert>
      </Snackbar>
    </SnackbarContext.Provider>
  );
};

export const useSnackbar = () => {
  const context = useContext(SnackbarContext);
  if (context === undefined) {
    throw new Error('useSnackbar must be used within a SnackbarProvider');
  }
  return context;
};
