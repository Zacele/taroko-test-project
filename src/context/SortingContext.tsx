'use client';
import { SortingOrderType } from '@/interface';
import React, { createContext, useContext, useState, ReactNode } from 'react';

interface SortingContextType {
  order: SortingOrderType;
  setOrder: (order: SortingOrderType) => void;
}

const SortingContext = createContext<SortingContextType | undefined>(undefined);

interface SortingProviderProps {
  children: ReactNode;
}

export const SortingProvider: React.FC<SortingProviderProps> = ({
  children,
}) => {
  const [order, setOrder] = useState<SortingOrderType>('desc');

  return (
    <SortingContext.Provider value={{ order, setOrder }}>
      {children}
    </SortingContext.Provider>
  );
};

export const useSortingContext = () => {
  const context = useContext(SortingContext);
  if (context === undefined) {
    throw new Error('useSortingContext must be used within a SortingProvider');
  }
  return context;
};
