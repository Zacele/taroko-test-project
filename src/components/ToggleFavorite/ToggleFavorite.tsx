'use client';
import React from 'react';
import styles from './ToggleFavorite.module.css';
import clsx from 'clsx';
import { useFavorites } from '@/context/FavoritesContext';

const ToggleFavorite: React.FC = () => {
  const [showFavorites, setShowFavorites] = React.useState(false);

  const { setFavoritesFilterOn } = useFavorites();

  const onToggleFavoritesClick = () => {
    setShowFavorites((prev) => !prev);
    setFavoritesFilterOn(!showFavorites);
  };

  return (
    <button
      onClick={onToggleFavoritesClick}
      className={clsx(styles.button, {
        [styles.isButtonHighlight]: showFavorites,
      })}
    >
      <span
        className={clsx(styles.star, {
          [styles.isButtonHighlightSpan]: showFavorites,
        })}
      >
        â˜…
      </span>
      Favorites
    </button>
  );
};

export default ToggleFavorite;
