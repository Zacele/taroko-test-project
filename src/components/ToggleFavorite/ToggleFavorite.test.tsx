import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import ToggleFavorite from './ToggleFavorite';
import { useFavorites } from '@/context/FavoritesContext';

// Mock the FavoritesContext
jest.mock('@/context/FavoritesContext', () => ({
  useFavorites: jest.fn(),
}));

describe('ToggleFavorite', () => {
  beforeEach(() => {
    // Clear all mocks before each test
    jest.clearAllMocks();

    // Setup mock return value for useFavorites
    (useFavorites as jest.Mock).mockReturnValue({
      setFavoritesFilterOn: jest.fn(),
    });
  });

  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(<ToggleFavorite />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('toggles favorites filter on button click', async () => {
    const setFavoritesFilterOnMock = jest.fn();
    (useFavorites as jest.Mock).mockReturnValue({
      setFavoritesFilterOn: setFavoritesFilterOnMock,
    });

    render(<ToggleFavorite />);
    const buttonElement = screen.getByRole('button', { name: /favorites/i });
    await userEvent.click(buttonElement);

    expect(setFavoritesFilterOnMock).toHaveBeenCalledWith(true);
    await userEvent.click(buttonElement);
    expect(setFavoritesFilterOnMock).toHaveBeenCalledWith(false);
  });
});
