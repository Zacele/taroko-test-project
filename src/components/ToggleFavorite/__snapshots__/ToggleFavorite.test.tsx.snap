// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleFavorite renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="button"
  >
    <span
      class="star"
    >
      â˜…
    </span>
    Favorites
  </button>
</DocumentFragment>
`;
